File: Chains/Multishine.cpp
#include "Multishine.h"
#include "../Util/Logger.h"

Multishine::Multishine()
{
    m_startingFrame = m_state->m_memory->frame;
    Logger::Instance()->Log(INFO, "Start Multishine");
    m_canInterrupt = true;
}

Multishine::~Multishine()
{
}

bool Multishine::IsInterruptible()
{
    //Safety return
    uint frame = m_state->m_memory->frame - m_startingFrame;
    if(frame >= 30)
    {
        return true;
    }

    return m_canInterrupt;
}

void Multishine::PressButtons()
{
    m_canInterrupt = true;

    //If standing, shine
    if(m_state->m_memory->player_two_action == STANDING)
    {
        m_controller->pressButton(Controller::BUTTON_B);
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, 0);
        m_canInterrupt = false;
        return;
    }

    //Shine on the last frame of knee bend
    if(m_state->m_memory->player_two_action == KNEE_BEND &&
        m_state->m_memory->player_two_action_frame >= 3)
    {
        m_controller->pressButton(Controller::BUTTON_B);
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, 0);
        m_canInterrupt = false;
        return;
    }

    if(m_state->m_memory->player_two_action == KNEE_BEND)
    {
        m_controller->emptyInput();
        return;
    }

    //Jump on the last frame of shine stun
    if((m_state->m_memory->player_two_action == DOWN_B_STUN ||
        m_state->m_memory->player_two_action == DOWN_B_GROUND_START) &&
        m_state->m_memory->player_two_action_frame >= 4 &&
        m_state->m_memory->player_two_on_ground)
    {
        m_controller->pressButton(Controller::BUTTON_Y);
        return;
    }

    if(m_state->m_memory->player_two_action == DOWN_B_GROUND)
    {
        m_controller->pressButton(Controller::BUTTON_Y);
        return;
    }

    if(m_state->m_memory->player_two_action == DOWN_B_STUN)
    {
        m_controller->emptyInput();
        return;
    }

    m_controller->emptyInput();
}

File: chains/multishine.py
from chains import chain
from util import enums
import globals

class MultiShine(chain.Chain):

    def pressbuttons(self):
        game_state = globals.gamestate
        smashbot_state = globals.smashbot_state
        opponent_state = globals.opponent_state
        controller = globals.controller

        #If standing, shine, become uninteruptible
        if smashbot_state.action == enums.Action.STANDING:
            controller.press_button(enums.Button.BUTTON_B)
            controller.tilt_analog(enums.Button.BUTTON_MAIN, .5, 0)
            self.interruptible = False
            return

        #Shine on frame 3 of knee bend, else nothing
        if smashbot_state.action == enums.Action.KNEE_BEND:
            if smashbot_state.action_frame == 2:
                controller.press_button(enums.Button.BUTTON_B)
                controller.tilt_analog(enums.Button.BUTTON_MAIN, .5, 0)
                self.interruptible = False
                return
            else:
                controller.empty_input()
                return

        isInShineStart = (smashbot_state.action == enums.Action.DOWN_B_STUN or
                smashbot_state.action == enums.Action.DOWN_B_GROUND_START)

        #Jump out of shine
        if isInShineStart and smashbot_state.action_frame >= 4 and smashbot_state.on_ground:
            controller.press_button(enums.Button.BUTTON_Y)
            return

        if smashbot_state.action == enums.Action.DOWN_B_GROUND:
            controller.press_button(enums.Button.BUTTON_Y)
            return

        controller.empty_input()
